### ISSUE TITLE
How can I judge that a x3domNode is no longer used? 

### ISSUE THREAD

**Issue by microaaron** [x3dom/x3dom#1220]
The USE node is still valid after the DEF node is removed.
Like this: [https://jsitor.com/S7EVHXwFB](https://jsitor.com/S7EVHXwFB).
Shape1Node still works on Gp2 and Gp3 but its _nameSpace and _xmlNode have been emptied. It looks a little weird but maybe it matches the design.
This brings a new problem. How can I judge that an x3domNode is no longer used? Neither onRemove() nor parentRemoved() seem to be able to judge.

**Comment by microaaron** [comment]
I think it is necessary to judge the number of _parentNodes when executing parentRemoved(), if the number is zero, remove all childNodes.

https://jsitor.com/MbzdqNpih

```
parentRemoved : function ( parent )
{
          if(this._parentNodes.length == 0)
          {
              console.log(this._DEF+" is no longer used.");
              for(var child of this._childNodes)
              {
                  this.removeChild(child,"any",true);  //remove this node from child's _parentNodes then trigger child.parentRemoved()
              }
          }
}
```

**Comment by microaaron** [comment]
Yes, Gp2 and Gp3 still have their _x3domNode after the DEF node is removed.  "document.getElementById("Gp1").remove()" just cuts off the parent-child relationship between SCENE and Gp1 by removeChild(). 

Removing USE nodes may also trigger onRemove() and parentRemoved(). OnRemove() and parentRemoved()don't know which DOM node triggered them, nor can they see the relationship between DOM nodes. They can only handle relationships between x3d nodes.

**Comment by andreasplesch** [comment]
I think in X3D removing a USE node in the scene graph means actually removing the node everywhere because there is only one node (the original DEF node which can have multiple parents). It may make sense to look into the SAI documentation/specification, and ask on x3d-public.

**Comment by microaaron** [comment]
I submit this issue because of a memory release problem. My solution is to remove all child nodes of an x3d node when its number of parents is zero, making this x3d node unreferenced.
As for how to automatically remove all DEF and USE nodes, I have no idea. Maybe it's not bad to let web programmers choose whether to delete them or not, even if it doesn't conform to the x3d specification

**Comment by microaaron** [comment]
Yes, when Gp1's DOM node is removed, Gp1's x3dom node just cuts off the parent-child relationship with SCENE, but it is still the parent of GP_1_1, and Gp2 is also the parent of GP_1_1, and Gp2's parent is SCENE. So Gp1 will not be released. I think since Gp1 has no parent, it should cut off the parent-child relationship with GP_1_1.

**Comment by microaaron** [comment]
Perhaps according to the standard, removing an x3d node does not mean removing its children, which can still be children of other nodes. But removing a DOM necessarily removes its children.

**Comment by andreasplesch** [comment]
It turns out I was wrong. Removing a DEF node does not remove it everywhere in the scene, in X3D. It removes only one occurrence.  Similarly, removing a USE node removes only one occurrence.

So it looks like x3dom is doing it correctly.

**Comment by microaaron** [comment]
Yes, I also think x3dom is doing it correctly. The next step is to deal with the memory release problem.

**Comment by andreasplesch** [comment]
In addition, X3D has a function to make a node unnamed, so it cannot be referenced anymore in the future. This does not affect existing occurrences. I do not think this function is currently implemented.

It also appears that a DEF node which became completely unreferenced and unlinked to a scene can still later be readded as a USE node. This means its composition cannot be deleted from browser memory (unless it also becomes unnamed).

**Comment by microaaron** [comment]
Currently, when a DEF is removed, its name is removed from the defMap.

**Comment by microaaron** [comment]
It may not be necessary to consider whether x3d nodes affect DOM nodes, just like setFieldValue doesn't affect DOM node, but setAttribute does.
